README.md
# campus-mails-manager
> To digitalize College of Wooster's mailing workflow.

---
A proper `README.md` file will be written once the program is fully developed. During the development stage, the records will be update on [`CHANGELOG.md`](https://github.com/cs230s19/campus_mails_manager/blob/master/CHANGELOG.md).


---
## Purpose of This Project
The purpose of this project is to digitalize College of Wooster's mailing workflow — mostly regarding delivery archiving, notifying, and redeeming aspects.

In shorts, rather than having the mailing room staffs manually sort the deliveries, fill out the blue slips, put them in the recipient's mail box, have the recipient to go and redeem it, then keep the redeemed blue slips for another month... We aim to design an app which can handle all above process digitally and in a more automated manner by providing features of:

* Digital registration of deliveries (even through a scan).
* Searchable database for registered deliveries.
* Automatic email notification to recipients.
* Redeem with QRcode.
* System generated logs.

## UI Showcase & User Manual.

![UI Demo](https://github.com/cs230s19/campus_mails_manager/blob/master/ui_demo.png)

* **Input Screen | Harry Dunham**
    * This screen is the default screen of the application, so the staff may quickly access it and input information regarding the just arrived package.

* **Log Screen | Henry Zhong**
    * This screen is to display all registered packages with filters available, so the staff may quickly find the desired log of the package if requested.

* **Details Screen | Nathan Devereux**
    * This screen is to show all information regarding a specific package entry, for the scenario of redeeming confirmation.



## Vision for Future Possibilities.

To actually this make program functional within the campus, we need a backend support which we don't know well enough yet. But in terms of the front end design, these are the future goals.

* **Input Screen**
    * Bar code scanner.
    * Time stamp `registered_time` input.
    * Profile for `staff_id`.

* **Log Screen**
    * More modular/customizable in terms of the information displayed for each entry
    * Embed `Spinner` or `DropDown List` for `carrier` and `redeemed_status`
    * Fuzzy match for searching.

* **Details Screen**
    * Better communication with other two screen.
    * Edit / Delete feature.
    * View multiple entries continuously (e.g. entries selected in `Details Container` from `Log Screen`)

## Acknowledgements.

* **Henry Zhong**
    * Responsible for developing `Log Screen`.
    * Responsible for designing the framework of the application.
    * Writing this document.

* **Harry Dunham**
    * Responsible for developing `Input Screen` with a fake progress bar.

* **Nathan Devereux**
    * Responsible for developing `Details Screen` and its linkage with `Log Screen`.


CHANGELOG.md
# CHANGELOG

> This file serves as the journal for the [campus_mails_manager](https://github.com/cs230s19/campus_mails_manager) project.

---

## Broadcasts

### 2019-04-10 | Final scrum meeting (17:00 @Andrews)
* Everyone should have his parts done by the time of meeting, so that the team can get together, merge three screens into one program and run some test trials.
* The team should polish this `CHANGELOG.md` file and update the [`README.md`](https://github.com/cs230s19/campus_mails_manager/blob/master/README.md).

### 2019-04-10 | Change on `registered.data`.

* Values of `tracking_num`, `box_num`, `bin_num` are now in `string` type.
    * I apologize for editing the most fundamental file at this stage, but I believe by having all values store as `string` would simplify the searching implementation (avoid type-check during looping). Also it is more consistent with the nature of all input features, as `TextInput.text` is default to be `string`.
* Key `service` is now `carrier` for better description.
* Key `delivered_time` is now `registered_time` for better description, consistency with `registered_staff`, and not to be confused with the delivery time of the carrier if we will ever do the bar code stuff.
* `is_redeemed` is now `redeemed_status`. Also, rather than having a boolean value, `redeemed_status` now has alternative values of `"Redeemed"` or `"Unredeemed"` — this make the display output much easier.

### 2019-04-09 | Scrum meeting after class.
Everyone are expected to have a screen with fully functional UI and be able to interact with the `registered.data` by the meeting.

### 2019-04-07 | A short style guide for updating this `CHANGELOG.md`.
* The `CHANGELOG.md` has three sections.
    * **Broadcasts**: Briefing of upcoming events or reminders (like this one).
    * **Scrum Meetings**: Summaries for scrum meetings.
    * **Development Journal**: Records for development updates.
* For the sake of consistency, I suggest we update this `CHANGELOG.md` in a date descending order (new-to-old) with a format of `YYYY-MM-DD`.



---
## Scrum Meetings

### 2019-04-10
* Tried to access widget attributes from another screen in python, mostly failed. Thus we simplify the design for `Details Screen` with a bypass button on `Log Screen`.
* Merged all three screens together.
* Adopted the design of having three screen-shift buttons in the bottom bar.

### 2019-04-09
* Achieved demo UI for all three screens.
* The task of interacting with JSON must be postponed due to unforeseen amount of work.
* Discuss the possibility of editing `registered.data`.

### 2019-04-06
* Sketched and discussed the UI design for each screens.
* Emphasized the necessary buttons to navigate between screens.
    * `Input Screen` <-- `Submit Button` within bottom bar --> `Log Screen`
    * `Log Screen` -- `Edit Button` within each entries --> `Details Screen`
    * `Details Screen` -- `Back Button` --> `Log Screen`
* Specified duties of each `Screen`.
    * `Input Screen`
        * Collects info about the delivery and injects them into `registered.data`.
        * A visual cue indicating successful add of delivery entry.
    * `Log Screen`
        * Filters for entries.
        * Sorting options for entries.
        * A scrollable section to display registered entries from `registered.data`
    * `Details Screen`
        * Display the details of selected entry base on `registered.data`.
        * Edit the selected entry.
        * A `Back Button`.

### 2019-04-04
* Created [GitHub repo](https://github.com/cs230s19/campus_mails_manager) with necessary files.
* Updated the [`registered.data`](https://github.com/cs230s19/campus_mails_manager/blob/master/registered.data) with test trials in `JSON` format.

### 2019-04-02
* Discussed the general picture of the project, with screens distributed.
    * Harry: `Input Screen`
    * Henry: `Log Screen`
    * Nathan: `Details Screen`
* Set the output resolution to be a mimic of iPhone XS, as for `2436 * 1125`.



---
## Development Journal

### 2019-04-11, Henry Zhong:

* **Bug fixed on filter deselection**
    * When two or more filters are triggered, it is ineffective to deselect any perviously triggered filters — as for a deselected filter may still remain effective although the button has visually adjusted itself to `normal` state. Now the bug has been fixed.

### 2019-04-11, Henry Zhong:
* **Updated [`README.md`](https://github.com/cs230s19/campus_mails_manager/blob/master/README.md) for Class Presentation**
    * Purpose of this project
    * UI Showcase & User Manual.
    * Vision for future possibilities.
    * Acknowledgements.

### 2019-04-11, Henry Zhong:
**The Program is Functional for Demo Trails and Presentation**
* Merged the actual `Input Screen` and `Details Screen` into the framework.
* Finished some small adjustments on `Input Screen` code to be compatible with the recently the updated `registered.data` file and screen-jump bottom bar buttons.

### 2019-04-10, Nathan Devereux:
* **Fully implemented Details Screen.**
    * Added a button in the Log Screen that calls the function to get the details of a specific package.

### 2019-04-10, Henry Zhong:
**`LogScreen` is now fully functional**
* **JSON Interaction Implemented**
    * Being able to obtain JSON entries with generator, specifically:
        * Entries which with matched keys/values to the `search_dict`.
        * Entries with matched value to the unspecified `search_input`.
        * All entries.
* **Search Function Fully Implemented**
    * Color-coding implemented: including situation of select, deselect, and over-ruling select (selected a filter but did not submit an input, then selected another filter, the pervious filter will be deselect automatically).
    * Implemented wrapper functions to generate `search_dict` base on user input, thus the program can use such dictionary to interact with JSON.
    * Implemented some user friendly features:

log.py
from kivy.config import Config
Config.set('graphics', 'width', '563')
Config.set('graphics', 'height', '1218')


from kivy.app import App
from kivy.uix.togglebutton import ToggleButton
from kivy.uix.widget import Widget
from kivy.uix.label import Label
from kivy.uix.button import Button
from kivy.uix.popup import Popup
from kivy.uix.boxlayout import BoxLayout
from kivy.uix.gridlayout import GridLayout
from kivy.uix.spinner import Spinner
from kivy.clock import mainthread
from kivy.uix.screenmanager import Screen

import json, requests



def get_next_entry():
    with open('registered.data') as json_file:
        json_data = json.load(json_file)
        for entry in json_data:
            yield entry

search_dict = {}
def get_matched_entry():
    # app = App.get_running_app()
    # search_dict = app.root.ids.current_log_screen.__class__.search_dict
    for entry in get_next_entry():
        holder_list = []
        for search_key in search_dict:
            if entry[search_key]:
                if search_dict[search_key] == entry[search_key] or search_dict[search_key] in entry[search_key]:
                    holder_list.append(entry['tracking_num'])
        if len(holder_list) == len(search_dict):
            yield entry

def search_all_entry():
    target_content = search_dict['everything']
    search_dict.pop('everything', None)
    for entry in get_next_entry():
        for k, v in entry.viewitems():
            if v != False:
                if v == target_content or target_content in v:
                    yield entry
                    break

def entry_to_str(entry):
    entry_str = ('Tracking #: ' + entry['tracking_num'] +
                '  |  ' + entry['recipient'].center(20) +
                '  |  ' + entry['registered_time'] +
                '  |  ' + entry['redeemed_status'].center(15))
    return entry_str


def entry_to_display(mode):
    entry_str = ''
    if mode == 1:
        for entry in get_next_entry():
            entry_str = entry_to_str(entry)
            yield entry_str
    elif mode == 2:
        for entry in get_matched_entry():
            entry_str = entry_to_str(entry)
            yield entry_str
    elif mode == 3:
        for entry in search_all_entry():
            entry_str = entry_to_str(entry)
            yield entry_str



filters_dict = {'filter_tracking_num': False,
                'filter_recipient': False,
                'filter_service': False,
                'filter_is_redeemed': False,
                'filter_registered_time': False,
                'filter_redeemed_time': False,
                'filter_registered_staff': False,
                'filter_redeemed_staff': False}

def filter_format_LUT(filter_id_str):
    filter_format_LUT = {'filter_tracking_num': 'Enter the last four digits as "XXXX"',
                          'filter_recipient': 'Enter as "Firstname Lastname"',
                          'filter_carrier': 'Specify: "UPS" "DHL" "FEDEX" "USPOSTAL" "CAMPUS"',
                          'filter_redeemed_status': 'Specify: "Redeemed" "Unredeemed"',
                          'filter_registered_time': 'Enter as "YYYY-MM-DD"',
                          'filter_redeemed_time': 'Enter as "YYYY-MM-DD"',
                          'filter_registered_staff': 'Enter the stuff ID',
                          'filter_redeemed_staff': 'Enter the stuff ID'}
    return filter_format_LUT[filter_id_str]


class InputScreen(Screen):
    pass

class DetailsScreen(Screen):
    pass

class LogScreen(Screen):

    # search_dict = {}

    def __int__(self, **kwargs):
        super(LogScreen, self).__init__(**kwargs)
        # self.triggered_filter_str = ''



    def filter_trigger(self, filter_id_str):
        self.filter_id_LUT = {'filter_tracking_num': self.ids.filter_tracking_num,
                              'filter_recipient': self.ids.filter_recipient,
                              'filter_carrier': self.ids.filter_carrier,
                              'filter_redeemed_status': self.ids.filter_redeemed_status,
                              'filter_registered_time': self.ids.filter_registered_time,
                              'filter_redeemed_time': self.ids.filter_redeemed_time,
                              'filter_registered_staff': self.ids.filter_registered_staff,
                              'filter_redeemed_staff': self.ids.filter_redeemed_staff}

        self.triggered_filter_id = self.filter_id_LUT[filter_id_str]


        if self.triggered_filter_id.state == 'down':
            self.ids.search_input.hint_text = filter_format_LUT(filter_id_str)
            if self.ids.search_submit.search_scope:
                if self.ids.search_submit.search_scope not in search_dict:
                    self.last_triggered_filter_id = self.filter_id_LUT['filter_' + self.ids.search_submit.search_scope]
                    self.last_triggered_filter_id.state = 'normal'

            self.ids.search_submit.search_scope = filter_id_str[7:]
            print('Current search scope is: {}'.format(self.ids.search_submit.search_scope))

        elif self.triggered_filter_id.state == 'normal':
            self.ids.search_input.hint_text = 'Enter search content or select a filter for advanced search'
            if self.ids.search_submit.search_scope == filter_id_str[7:]:

                search_dict.pop(filter_id_str[7:], None)

                self.ids.search_submit.search_scope = False



    def submit(self):
        user_search_input = self.ids.search_input.text

        if self.ids.search_submit.search_scope != False:
            search_dict[self.ids.search_submit.search_scope] = user_search_input
            self.on_enter(2)
        else:
            search_dict['everything'] = user_search_input
            self.on_enter(3)
        print('Current search_dict is: {}'.format(search_dict))

    def reset(self):
        search_dict = {}
        self.ids.search_input.text = ''
        self.ids.search_submit.search_scope = False
        button_tuple = (self.ids.filter_tracking_num,
                        self.ids.filter_recipient,
                        self.ids.filter_carrier,
                        self.ids.filter_redeemed_status,
                        self.ids.filter_registered_time,
                        self.ids.filter_redeemed_time,
                        self.ids.filter_registered_staff,
                        self.ids.filter_redeemed_staff)
        for button in button_tuple:
            if button.state != 'normal':
                button.state = 'normal'
        self.on_enter(1)

    def to_details(self, instance):
        print('Proceed to the details page of #{}'.format(instance.text[12:16]))



    @mainthread
    def on_enter(self, mode=1):
        self.ids.entry_layout.clear_widgets()
        for entry_str in entry_to_display(mode):
            a_button = Button(text = entry_str)
            # a_spinner = Spinner(text=entry_str,
            #                     values = ('Edit #'+entry_str[12:16], 'Delete #'+entry_str[12:16]),
            #                     sync_height = True)
            # self.ids.entry_layout.add_widget(a_spinner)
            a_button.bind(on_release = self.to_details)
            self.ids.entry_layout.add_widget(a_button)




class logApp(App):
    pass

if __name__ == '__main__':
    logApp().run()

input.py
import json
from kivy.app import App
from kivy.uix.screenmanager import Screen, ScreenManager
from kivy.uix.boxlayout import BoxLayout
from kivy.uix.button import Button
from kivy.uix.popup import Popup
from kivy.uix.label import Label
from kivy.config import Config
from datetime import datetime
# datetime.utcnow()
# Config.set('graphics', 'width', '323')
# Config.set('graphics', 'height', '700')


class InputScreen(Screen):
    def submission_func(self):
        self.ids.status.value = 0
        selected_carrier = ''
        carrier_list = [self.ids.USPS, self.ids.Fedex, self.ids.UPS,
                        self.ids.DHL, self.ids.Other]
        for carrier in carrier_list:
            if carrier.state == 'down':
                selected_carrier = carrier.text

        if selected_carrier == '':
            self.popup()
            return
        information = {'tracking_num': self.ids.tracking_num.text, 'recipient':
                       self.ids.recipient.text, 'note': self.ids.note.text,
                       'bin_number': self.ids.bin_number.text, 'box_number':
                       self.ids.box_number.text, 'registered_staff':
                       self.ids.registered_staff.text, 'delivered_time':
                       self.ids.delivered_time.text, 'is_redeemed': 'false',
                       'redeemed_time': 'false', 'service': selected_carrier}

        for json_name, text in information.items():
            if text == '':
                self.popup()
                return

        filename = "registered.data"
        with open(filename, "r") as copy_file:
            try:
                data = json.load(copy_file)
            except ValueError:
                data = []

            data.append(information)
        with open(filename, "w") as json_file:
            json.dump(data, json_file)
            json_file.close()

        self.ids.tracking_num.text = ''
        self.ids.recipient.text = ''
        self.ids.note.text = ''
        self.ids.bin_number.text = ''
        self.ids.box_number.text = ''
        self.ids.registered_staff.text = ''
        self.ids.delivered_time.text = ''


        for carrier in carrier_list:
            carrier.state = 'normal'


    def popup(self):
        """
        This function implements popup widget functionality, whereby should an
        exception be thrown, a popup will be created to alert the user the
        error. The user then must acknowledge the popup by clicking the 'Ok'
        button. This physical instantiation of this includes a general
        BoxLayout for the entire popup, a Label widget to display the error
        message, and an 'Ok' button to accept the error.

        :param error_message: The exceptions error message, which will be
        displayed by the popup.
        :return: None
        """
        content = BoxLayout(orientation='vertical')
        error_label = Label(text="Please fill all required fields")
        dismiss_button = Button(text='Ok')
        content.add_widget(error_label)
        content.add_widget(dismiss_button)
        popup = Popup(title='Error', content=content, size_hint=(.3, .25),
                      auto_dismiss=False)
        dismiss_button.bind(on_press=popup.dismiss)
        popup.open()




class InputApp(App):
    def build(self):
        return Input()


if __name__ == '__main__':
    InputApp().run()




details.py
from kivy.app import App
from kivy.uix.screenmanager import Screen
import json




def get_entry(num):
    with open('registered.data') as json_file:
        json_data = json.load(json_file)
        for entry in json_data:
            if entry['tracking_num'] == num:
                return entry

num = '1234'
# app = App.get_running_app()
# app.root.ids.current_layout.last_entry()



class DetailsScreen(Screen):
    #
    # def __init__(self):
    #     super(DetailsScreen, self).__init__(**kwargs)
    #
    # # def set_label_text(self):
    #
    #
    def on_enter(self):
        self.ids.output_label.text = 'something'
        # entry = get_entry(num)




    def process_submit(self):
        json_list = ['tracking_num', 'recipient', 'box_num',
                     'registered_staff', 'registered_time', 'bin_num',
                     'note', 'redeemed_staff', 'redeemed time', 'carrier']
        for i in range(0, 10):
            if self.ids.spin.values == self.ids.spin.values[i]:
                entry = get_entry('1234')
                entry[json_list[i]] = self.ids.edit.text
        self.last_entry(num)

    def process_redeem(self):
        entry = get_entry(num)
        entry['is_redeemed'] = 'false'


class detailsApp(App):
    pass


if __name__ == '__main__':
    detailsApp().run()


manager.py
from kivy.config import Config
Config.set('graphics', 'width', '563')
Config.set('graphics', 'height', '1218')


from kivy.app import App
from kivy.uix.togglebutton import ToggleButton
from kivy.uix.widget import Widget
from kivy.uix.label import Label
from kivy.uix.button import Button
from kivy.uix.popup import Popup
from kivy.uix.boxlayout import BoxLayout
from kivy.uix.gridlayout import GridLayout
from kivy.uix.spinner import Spinner
from kivy.clock import mainthread
from kivy.uix.screenmanager import ScreenManager, Screen
from kivy.properties import ObjectProperty
import json, requests
from datetime import datetime








class InputScreen(Screen):
    def submission_func(self):
        self.ids.status.value = 0
        selected_carrier = ''
        carrier_list = [self.ids.USPS, self.ids.Fedex, self.ids.UPS,
                        self.ids.DHL, self.ids.CAMPUS]
        for carrier in carrier_list:
            if carrier.state == 'down':
                selected_carrier = carrier.text

        if selected_carrier == '':
            self.popup()
            return
        information = {'tracking_num': self.ids.tracking_num.text,
                        'recipient': self.ids.recipient.text,
                        'box_num': self.ids.box_number.text,
                        'registered_staff': self.ids.registered_staff.text,
                        'registered_time': self.ids.delivered_time.text,
                        'bin_num': self.ids.bin_number.text,
                        'note': self.ids.note.text,
                        'redeemed_status': 'Unredeemed',
                        'redeemed_staff': 'Unknown',
                        'redeemed_time': 'Unknown',
                        'carrier': selected_carrier}

        for json_name, text in information.items():
            if text == '':
                self.popup()
                return

        filename = "registered.data"
        with open(filename, "r") as copy_file:
            try:
                data = json.load(copy_file)
            except ValueError:
                data = []

            data.append(information)
        with open(filename, "w") as json_file:
            json.dump(data, json_file)
            json_file.close()

        self.ids.tracking_num.text = ''
        self.ids.recipient.text = ''
        self.ids.note.text = ''
        self.ids.bin_number.text = ''
        self.ids.box_number.text = ''
        self.ids.registered_staff.text = ''
        self.ids.delivered_time.text = ''


        for carrier in carrier_list:
            carrier.state = 'normal'


    def popup(self):
        """
        This function implements popup widget functionality, whereby should an
        exception be thrown, a popup will be created to alert the user the
        error. The user then must acknowledge the popup by clicking the 'Ok'
        button. This physical instantiation of this includes a general
        BoxLayout for the entire popup, a Label widget to display the error
        message, and an 'Ok' button to accept the error.

        :param error_message: The exceptions error message, which will be
        displayed by the popup.
        :return: None
        """
        content = BoxLayout(orientation='vertical')
        error_label = Label(text="Please fill all required fields")
        dismiss_button = Button(text='Ok')
        content.add_widget(error_label)
        content.add_widget(dismiss_button)
        popup = Popup(title='Error', content=content, size_hint=(.3, .25),
                      auto_dismiss=False)
        dismiss_button.bind(on_press=popup.dismiss)
        popup.open()


















def get_next_entry():
    with open('registered.data') as json_file:
        json_data = json.load(json_file)
        for entry in json_data:
            yield entry

search_dict = {}
def get_matched_entry():
    # app = App.get_running_app()
    # search_dict = app.root.ids.current_log_screen.__class__.search_dict
    for entry in get_next_entry():
        holder_list = []
        for search_key in search_dict:
            if entry[search_key]:
                if search_dict[search_key] == entry[search_key] or search_dict[search_key] in entry[search_key]:
                    holder_list.append(entry['tracking_num'])
        if len(holder_list) == len(search_dict):
            yield entry

def search_all_entry():
    target_content = search_dict['everything']
    search_dict.pop('everything', None)
    for entry in get_next_entry():
        for k, v in entry.viewitems():
            if v != False:
                if v == target_content or target_content in v:
                    yield entry
                    break

def entry_to_str(entry):
    entry_str = ('Tracking #: ' + entry['tracking_num'] +
                '  |  ' + entry['recipient'].center(20) +
                '  |  ' + entry['registered_time'] +
                '  |  ' + entry['redeemed_status'].center(15))
    return entry_str


def entry_to_display(mode):
    entry_str = ''
    if mode == 1:
        for entry in get_next_entry():
            entry_str = entry_to_str(entry)
            yield entry_str
    elif mode == 2:
        for entry in get_matched_entry():
            entry_str = entry_to_str(entry)
            yield entry_str
    elif mode == 3:
        for entry in search_all_entry():
            entry_str = entry_to_str(entry)
            yield entry_str



filters_dict = {'filter_tracking_num': False,
                'filter_recipient': False,
                'filter_service': False,
                'filter_is_redeemed': False,
                'filter_registered_time': False,
                'filter_redeemed_time': False,
                'filter_registered_staff': False,
                'filter_redeemed_staff': False}

def filter_format_LUT(filter_id_str):
    filter_format_LUT = {'filter_tracking_num': 'Enter the last four digits as "XXXX"',
                          'filter_recipient': 'Enter as "Firstname Lastname"',
                          'filter_carrier': 'Specify: "UPS" "DHL" "FEDEX" "USPOSTAL" "CAMPUS"',
                          'filter_redeemed_status': 'Specify: "Redeemed" "Unredeemed"',
                          'filter_registered_time': 'Enter as "YYYY-MM-DD"',
                          'filter_redeemed_time': 'Enter as "YYYY-MM-DD"',
                          'filter_registered_staff': 'Enter the stuff ID',
                          'filter_redeemed_staff': 'Enter the stuff ID'}
    return filter_format_LUT[filter_id_str]



class LogScreen(Screen):


    def __int__(self, **kwargs):
        super(LogScreen, self).__init__(**kwargs)
        # self.triggered_filter_str = ''


    def filter_trigger(self, filter_id_str):
        self.filter_id_LUT = {'filter_tracking_num': self.ids.filter_tracking_num,
                              'filter_recipient': self.ids.filter_recipient,
                              'filter_carrier': self.ids.filter_carrier,
                              'filter_redeemed_status': self.ids.filter_redeemed_status,
                              'filter_registered_time': self.ids.filter_registered_time,
                              'filter_redeemed_time': self.ids.filter_redeemed_time,
                              'filter_registered_staff': self.ids.filter_registered_staff,
                              'filter_redeemed_staff': self.ids.filter_redeemed_staff}

        self.triggered_filter_id = self.filter_id_LUT[filter_id_str]


        if self.triggered_filter_id.state == 'down':
            self.ids.search_input.hint_text = filter_format_LUT(filter_id_str)
            if self.ids.search_submit.search_scope:
                print('first lay')
                if self.ids.search_submit.search_scope not in search_dict:
                    print('first lay')
                    self.last_triggered_filter_id = self.filter_id_LUT['filter_' + self.ids.search_submit.search_scope]
                    self.last_triggered_filter_id.state = 'normal'
                    self.ids.search_submit.search_scope

            print('outer lay')
            self.ids.search_submit.search_scope = filter_id_str[7:]
            print('Current search scope is: {}'.format(self.ids.search_submit.search_scope))

        elif self.triggered_filter_id.state == 'normal':
            self.ids.search_input.hint_text = 'Enter search content or select a filter for advanced search'
            # if self.ids.search_submit.search_scope == filter_id_str[7:]:

            search_dict.pop(filter_id_str[7:], None)

            self.ids.search_submit.search_scope = False



    def submit(self):
        user_search_input = self.ids.search_input.text

        # if self.ids.search_submit.search_scope != False:
        if self.ids.search_submit.search_scope != False:
            search_dict[self.ids.search_submit.search_scope] = user_search_input
            self.on_enter(2)
        elif len(search_dict) >= 1:
            self.on_enter(2)
        else:
            search_dict['everything'] = user_search_input
            self.on_enter(3)
        print('Current search_dict is: {}'.format(search_dict))

    def reset(self):
        search_dict.clear()
        self.ids.search_input.text = ''
        self.ids.search_submit.search_scope = False
        button_tuple = (self.ids.filter_tracking_num,
                        self.ids.filter_recipient,
                        self.ids.filter_carrier,
                        self.ids.filter_redeemed_status,
                        self.ids.filter_registered_time,
                        self.ids.filter_redeemed_time,
                        self.ids.filter_registered_staff,
                        self.ids.filter_redeemed_staff)
        for button in button_tuple:
            if button.state != 'normal':
                button.state = 'normal'
        self.on_enter(1)

    def entry_to_detail_str(self, entry):
        output_str = ''
        for k, v in entry.items():
            if k == 'tracking_num':
                continue
            output_str += (k + ': \n'
                            + '                                                '
                            + str(v) + '\n\n')
        return output_str


    def to_details(self, instance):
        print('Proceed to the details page of #{}'.format(instance.text[12:16]))
        entry_holder = False
        for entry in get_next_entry():
            if entry['tracking_num'] == instance.text[12:16]:
                entry_holder = entry
                break
        output_str = ('------------------------------------------------------------------------------------\n'
                        + 'Tracking #'+instance.text[12:16]
                        + '\n\n\n\n'
                        + self.entry_to_detail_str(entry)
                        + '------------------------------------------------------------------------------------\n')
        self.ids.details_trigger_button.hinding_text = output_str.center(25)
        self.ids.details_trigger_button.text = 'View #'+instance.text[12:16]+' in Details'






    @mainthread
    def on_enter(self, mode=1):
        self.ids.entry_layout.clear_widgets()
        for entry_str in entry_to_display(mode):
            a_button = Button(text = entry_str)
            a_button.bind(on_release = self.to_details)
            self.ids.entry_layout.add_widget(a_button)




class DetailsScreen(Screen):
    pass


class managerApp(App):
    pass

if __name__ == '__main__':
    managerApp().run()